src/bn256/mod.rs

```diff
@@ -69,15 +71,32 @@ impl Bn256 {
         let mut c = 0..255;
 
         // Add counter suffix
+        // This message should be: ciphersuite || 0x01 || message || ctr
+        // For the moment we work with message || ctr until a tag is decided
         let mut v = [&message[..], &[0x00]].concat();
         let position = v.len() - 1;
 
-        // `Hash(cipher||PK||data)`
+        // `Hash(data||ctr)`
+        // The modulus of bn256 is low enough to trigger several iterations of this loop
+        // We instead compute attempted_hash = `Hash(data||ctr)` mod Fq::modulus
+        // This should trigger less iterations of the loop
         let point = c.find_map(|ctr| {
             v[position] = ctr;
-            let attempted_hash = self.calculate_sha256(&v);
-            // Check validity of `H` (i.e. point exists in group G1)
-            self.arbitrary_string_to_g1(&attempted_hash).ok()
+            let hash = &self.calculate_sha256(&v)[0..32];
+            // this should never fail as the length of sha256 is max 256
+            let mut attempted_hash = arith::U256::from_slice(hash).unwrap();
+
+            // the library does not provide a function to do a modulus
+            // we use the provided add function adding a 0
+            // we also need to iterate here as the library does the modulus only once
+            while attempted_hash > Fq::modulus() {
```

Comment:
Here we should check for modulo bias, if `attempted_hash >= last_multiple_of_fq_modulus_lower_than_2_256` we should retry the hash with the next counter value.

Reply:
fixing it